RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,509 @@
+    torch.sparse_coo_tensor(indices, values, size=None, fill_value=None, dtype=None, device=None, requires_grad=False)
+    ```
+
+2.  The default fill-value is `0`.
+
+    This choice is consistent with the interpretation of unspecified
+    element in Linear Algebra domain.
+
+3.  PyTorch functions that have `layout` argument, may use different
+    fill-value when constructing a sparse tensor as defined by the
+    following table:
+
+    | Function              | Fill-value of returned sparse tensor |
+    | :-------------------- | :----------------------------------- |
+    | `torch.arange`        | 0                                    |
+    | `torch.empty`         | 0                                    |
+    | `torch.empty_like`    | 0                                    |
+    | `torch.empty_strided` | N/A                                  |
+    | `torch.eye`           | 0                                    |
+    | `torch.full`          | same as `fill_value` argument        |
+    | `torch.full_like`     | same as `fill_value` argument        |
+    | `torch.linspace`      | 0                                    |
+    | `torch.logspace`      | 0                                    |
+    | `torch.ones`          | 1                                    |
+    | `torch.ones_like`     | 1                                    |
+    | `torch.range`         | 0                                    |
+    | `torch.zeros`         | 0                                    |
+    | `torch.zeros_like`    | 0                                    |
+
+    For discussion: should we introduce the `fill_value` argument to
+    functions `empty` and `empty_like`?
+
+4.  The fill-value of a sparse tensor can be acquired via
+    `fill_value()` method that returns a strided `torch.Tensor`
+    instance that has the same dtype and storage location properties
+    as the elements of the sparse tensor:
+
+    - `A.fill_value().dtype == A.dtype`
+    - `A.fill_value().device == A.device`
+
+    The fill-value of a sparse non-hybrid tensor is a scalar tensor
+    `torch.tensor(0, dtype=A.dtype, device=A.device)`.
+
+5.  The fill-value of a hybrid sparse tensor has the same shape as the
+    dense part of the tensor.
+
+    For instance, for a sparse tensor `A` in COO storage format we
+    have:
+
+    - `A.fill_value().shape == A.values().shape[1:]`
+
+    While an obvious alternative to tensor fill-value would be a
+    scalar fill-value, the scalar fill-value would complicate the
+    implementation of algorithms for hybrid tensors. For
+    instance:
+
+    1. Accessing a dense element of a hybrid sparse tensor, the result
+       is a dense tensor. When this element corresponds to unspecified
+       element, one needs to construct a new tensor and fill it with
+       the scalar fill-value. A more efficient and simple way is to
+       just return the output of `fill_value()` that does would not
+       involve constructing a new tensor on each method call.
```

Comment:
The "more memory" argument is not really a problem considering that the `fill_value` would consume the same amount of memory that a single element of a sparse tensor does. I think we can assume that in a typical application of sparse tensors, the number of specified elements is far greater than 1.

For example, consider a 1000x1000 matrix with 10000 specified elements. Optimizing for memory would give saving 1/10000 of total memory usage for this particular example. However, if one would loop, say, over all elements of the tensor, there would be 1000000 minus 10000 newly created tensor instances corresponding to all unspecified elements. While the second example is somewhat extreme, I would give more weight to not creating a bunch of new tensor instances than saving 0.01 % of memory.

Reply:
I'm thinking there are sparse tensors that are consisting of _only_ dense blocks (e.g. to represent block-sparse type structure), and for those types of tensors memory use would double.

Am I misunderstanding something here about the hybrid format?

Reply:
Oh wait, okay - my assumption here about what you were proposing was wrong. Agreed, this seems like a very minor concern.

Reply:
PyTorch hybrid tensor is a sparse tensor in COO format, that is, it has `indices` and `values` to store the tensor data, with the difference that the size of `values` elements is not 0 (that would be a scalar and correspond to traditional COO tensor) but can have any dimension. `values` is a strided tensor. See the invariants in https://pytorch.org/docs/stable/sparse.html

So, no, hybrid tensor does not represent a block-sparse type of matrix where the matrix consists of rectangular blocks that are dense and anything in between are zeros. Hybrid tensor can represent certain types of block-sparse matrices: the blocks would need to be full rows, for instance.

Introducing tensor fill-value to hybrid tensors would certainly not double the memory usage. The increase of memory usage would be equal to the memory usage of one element of the sparse part of the hybrid tensor.

Reply:
See https://ars.els-cdn.com/content/image/1-s2.0-S0743731518305161-gr1.jpg where
- the first image represents a traditional 3D sparse tensor
- the second image represents a 3D hybrid tensor with 2 sparse dimensions and 1 dense dimension.
- the last image represents a 3D tensor with 1 sparse dimension and 2 dense dimensions

Reply:
One can achieve the benefits of having both a dense and a scalar by saying simply that the fill-value should be broadcast-able to the dense shape of the tensor, instead of restricting it to just be the dense shape.

Reply:
@hameerabbasi I like this idea a lot because it would resolve the zero fill-value test performance issue.
For element-wise operations, the idea works. For row-wise operations (like softmax), the idea needs to be analyzed because there the fill-value would need to be resized from scalar to 1D tensor in-place.

I'll unresolve the conversation to address @hameerabbasi point only.

Reply:
Yes, in the case of softmax it would need a new allocation, or else you would need a way of looping over strided fill-values within sparse kernels. The latter would also resolve your issue.

Reply:
According to updated proposal, one can specify hybrid tensor fill value that shape can be smaller that the shape of tensor dense elements. The specified fill value is available via `_fill_value()` method while the fill value of hybrid tensor can be acquired via `fill_value()`. Element-wise ops are applied to `_fill_value()` and this is also used for zero fill value testing. This approach resolves the performance issue optimally.